Defer expensive layout calcs for InfiniteScroll and StickyPanel.

sgomes opened the pull request:
InfiniteScroll and StickyPanel perform expensive layout calculations on start, but they aren't actually needed until a scroll occurs.

Deferring the calculations allows them to happen after the initial event loop and thus the initial render, improving responsiveness.

These changes massively improve the event response time for navigating to the Themes page. The first time is affected by other factors, but the event response time on subsequent navigations is now ~500ms instead of ~1200ms on my machine.

sgomes committed:
Defer expensive layout calcs for InfiniteScroll and StickyPanel.

InfiniteScroll and StickyPanel perform expensive layout calculations
on start, but they aren't actually needed until a scroll occurs.

Deferring the calculations allows them to happen after the initial
event loop and thus the initial render, improving responsiveness.

These changes massively improve the load time for the Themes page.
The first load is affected by other factors, but subsequent loads
now take ~500ms instead of ~1200ms on my machine.

jsnajdr commented on client/components/infinite-scroll/index.js:
```diff
@@ -20,12 +20,17 @@ export default class InfiniteScroll extends Component {
 	componentDidUpdate() {
-		this.checkScrollPosition( false );
+		if ( this.initialMount ) {
+			this.initialMount = false;
```
What is the rationale behind this check? When does it happen that `componentDidMount` is immediately followed by `componentDidUpdate`?

Could the `checkScrollPosition` in `componentDidUpdate`  also be deferred? There can be plenty of `componentDidUpdate`s in rapid succession when components rerender often.

Would it be useful if the used `throttle` or `debounce` to defer the `checkScrollPosition` calls and rate-limit them on frequent updates? That's just an idea.

sgomes replied:
I seem to have misinterpreted the React component lifecycle. There must be a different reason why I was getting a `componentDidUpdate` after the `componentDidMount`, perhaps as a result of one of the other issues you're mentioning.

I'll investigate further, and look into throttling or debouncing `checkScrollPosition` too.

jsnajdr commented:
Here's a funny stacktrace I found when testing the patched `InfiniteScroll`:

<img width="687" alt="screen shot 2018-07-24 at 11 21 51" src="https://user-images.githubusercontent.com/664258/43130168-5de7443a-8f36-11e8-8bdc-81b8f81b57c2.png">

`checkScrollPosition` is called recursively and causes a synchronous layout flush (:-1:):
1. User scrolls down and `checkScrollPosition` finds out it should load next page of results
2. `fetchNextPage` calls `incrementPage` which calls `setState` and the `ThemesSelection` component is rerendered (synchronously).
3. After the rerender, `componentDidUpdate` is called, which calls `checkScrollPosition` again.
4. Because i) this call happens at a time when DOM was already modified by the rerender, and i) the animation frame is not flushed yet, and iii) `checkScrollPosition` reads from DOM, layout is flushed synchronously to have the up-to-date box sizes available!

Is there a way to avoid the synchronous layout here?

Also, please have a look at `lib/after-layout-flush`. It's a helper to schedule callback after the next layout flush, when it's safe and efficient to read from DOM. Maybe some of your optimizations could use it.

`lib/after-layout-flush` should be used for sure by the `StickyPanel.onWindowScroll`. The `updateIsSticky` callback reads stuff from DOM, but `requestAnimationFrame` schedules it **before** the layout flush and almost guarantees synchronous layout! The original author probably didn't fully understand what `requestAnimationFrame` does.

blowery commented on client/components/infinite-scroll/index.js:
```diff
@@ -20,12 +20,17 @@ export default class InfiniteScroll extends Component {
+		this.initialMount = true;
 		window.addEventListener( 'scroll', this.checkScrollPositionHandler );
-		this.checkScrollPosition( false );
+		defer( () => this.checkScrollPosition( false ) );
```
The component could unmount before the scroll position check runs. We should either stash the timer handle and cancel it in componentWillUnmount

jsnajdr replied:
`defer` returns a timer ID that can be passed to `clearTimeout`.

sgomes replied:
That's a very good point, and it applies to both components. Thanks! I'll make that change.

blowery commented:
Thanks for the patch @sgomes! That's a nice enhancement. I left one issue that should make this able to land.

sgomes commented:
Excellent analysis on the component, @jsnajdr! Looks like I should have gone a bit deeper on `InfiniteList`, instead of just applying a small patch on top.

I'll make the change @blowery suggested first, and then I'll look a bit more holistically at what's happening on `InfiniteList`.

sgomes committed:
Ensure deferred actions in InfiniteScroll and StickyPanel get cleared

sgomes commented:
Thanks for the comments, @alisterscott, @blowery and @jsnajdr!

I'm taking the opportunity to do some work on this component, since @jsnajdr identified some serious issues in how it currently works.

My idea was to do an `IntersectionObserver`-based implementation as the main path, with the current scroll event-based implementation as fallback, after some heavy throttling and better analysis of the event flow. It will be a breaking change for consumers of the component (and I'll update them as needed), but I think it'll be worth having a codepath with no synchronous layout calls.

Would you prefer if I remove all changes to `InfiniteScroll` from this PR and take care of that work in a separate one?

sgomes committed:
Add IntersectionObserver-based implementation to InfiniteScroll.

Also reworks the scroll event-based implementation in order to
avoid loading and potential scheduling issues.

sgomes committed:
Use afterLayoutFlush with StickyPanel

sgomes commented:
Hello again, @alisterscott, @blowery and @jsnajdr!

I went ahead and reimplemented `InfiniteScroll` as part of this PR, since the changes weren't that numerous. It now uses `IntersectionObserver` where available, with a scroll-based fallback everywhere else.

I've also used `afterLayoutFlush` in `StickyPanel`, as @jsnajdr suggested.

This should be ready for another review, so please take another look! :)

blowery commented:
@sgomes thanks! we'll take a look.

sgomes commented:
Rebased on top of master.

sgomes commented:
@jsnajdr Any chance of a re-review? This page is the worst offender in terms of responsiveness as of right now, and this PR really makes a difference.

blowery commented:
Seems good to me. @jsnajdr see anything odd here?

jsnajdr commented:
@blowery There are a few things I want to review and test before merging.

jsnajdr commented on client/components/infinite-scroll/index.js:
```diff
@@ -3,53 +3,92 @@
 
 	componentDidUpdate() {
-		this.checkScrollPosition( false );
+		if ( ! this.hasIntersectionObserver && ! this.deferredTimer ) {
```
I think we lost one useful feature of `InfiniteScroll` here: if the page is not fully populated by current items, it fetches next page repeatedly until it gets populated and the page is long enough to scroll.

That's achieved by checking the scroll position in `componentDidUpdate`, which is called when a new page of results is rendered.

With `hasIntersectionObserver`, this check is never performed.

sgomes replied:
Yup, this was missing indeed.

I've reworked the logic around this, so it should now match the behaviour of the scroll event version; `componentDidUpdate` will now keep scheduling new fetches until the anchor element is no longer visible.

jsnajdr commented on client/components/infinite-scroll/index.js:
```diff
@@ -3,53 +3,92 @@
+	handleIntersection = entries => {
+		if ( entries && entries[ 0 ] && entries[ 0 ].isIntersecting ) {
+			const { boundingClientRect, rootBounds } = entries[ 0 ];
+			const triggeredByScroll = boundingClientRect.bottom >= rootBounds.bottom;
```
When I scroll down the `/themes` page (logged-out version), the `triggeredByScroll` param is always `false`, although the fetch was triggered by scroll.

We use the flag to send analytics about user behavior on the page.

jsnajdr replied:
I experimented with intersection observers a little and discovered that the callback is not called immediately and synchronously at the very moment the element scrolls into view. Therefore the exact equality check
```js
boundingClientRect.bottom === rootBounds.bottom
```
is not reliable at all. The element can be scrolled a few more pixels into the viewport at the time when the intersection observer reports the intersection.

The following condition:
```js
isIntersecting && boundingClientRect.bottom > rootBounds.bottom
```
can never be true for a zero-height element and/or threshold of 1.0. If the bottom edge of the observed element is below the root bottom edge, then it cannot be intersecting.

sgomes replied:
Yes, this logic was completely wrong, thank you for looking so deeply into this. I should have tested this more thoroughly.

I've completely reworked the logic here, by instead keeping track of when the anchor element first stops intersecting with the viewport. I think it's safe to assume that any fetches after that are triggered by scroll.

jsnajdr reviewed (commented):
One idea to consider: we have two implementations of the same component now. What if we separated them into two components and returned the right one after a feature detection?
```js
export default hasIntersectionObserver
  ? InfiniteScrollWithIntersectionObserver
  : InfiniteScrollWithScrollEvent;
```
That could make the implementation more readable. By avoiding a lot of `if`s.

sgomes commented:
Thanks for all the review comments, @jsnajdr! I'll start going through your suggested changes tomorrow. 👍

jsnajdr commented on client/components/infinite-scroll/index.js:
```diff
@@ -3,53 +3,92 @@
+			this.observer.observe( this.observedElement.current );
+		} else {
+			window.addEventListener( 'scroll', this.checkScrollPositionHandler );
+			this.deferredTimer = defer( () => this.checkScrollPositionHandler( false ) );
```
With `throttle` and `afterLayoutFlush`, this `defer` call becomes unneccessary. `afterLayoutFlush` itself schedules a rAF followed by `setTimeout( 0 )`. So, with `defer`, the chain becomes:
```js
setTimeout( // from defer
  () => requestAnimationFrame( // from afterLayoutFlush
    () => setTimeout( // fromAfterLayoutFlush
      checkScroll,
      0
    )
  ),
  0
);
```
See the extra `setTimeout( 0 )`? 

Fun fact: with `const throttled = throttle( check, 400 )`, the first call to `throttled` calls `check` synchronously, only the subsequent calls withing the throttling period are scheduled for later.

sgomes replied:
Thank you for clearing all of that up! Yes, it does seem I was over-scheduling. Fixed!

jsnajdr commented on client/components/infinite-scroll/index.js:
```diff
@@ -3,53 +3,92 @@
 	componentWillUnmount() {
+		// Clean up scroll codepath.
 		window.removeEventListener( 'scroll', this.checkScrollPositionHandler );
+		this.clearTimers();
```
After `this.clearTimers()`, there can still be dangling scheduled callbacks after unmounting. They are scheduled by `throttle` and `afterLayoutFlush` in `checkScrollPositionHandler`. Both wrappers return a function decorated with a `cancel` method that can be used to cancel them.

sgomes replied:
Very good point, we want to avoid that. Should be fixed now.

sgomes committed:
Split InfiniteScroll in two component versions.

One of them uses IntersectionObserver and the other scroll events.

sgomes committed:
Rework 'triggeredByScroll' login in InfiniteScroll.

Should now return true for loads after the initial set.
Also continuously fetches new pages until the page is scrollable.

sgomes committed:
Remove unnecessary defers from InfiniteScroll scroll impl.

Also cleans up timers properly.

sgomes commented:
Thank you again for the thorough review, @jsnajdr! I think I addressed all of the issues, so please take another look when you have the time.

Sorry for all the back-and-forth!

jsnajdr commented on client/components/infinite-scroll/index.js:
```diff
@@ -3,36 +3,118 @@
 	}
 
 	checkScrollPosition( triggeredByScroll ) {
+		this.clearPending();
```
This call is not needed at all. And it defeats the throttling: it will reset the timer inside `throttle` after each call.

sgomes replied:
Removed. Thank you!

jsnajdr commented on client/components/infinite-scroll/index.js:
```diff
@@ -3,36 +3,118 @@
+	}
+
+	checkScrollAfterLayout = () => {
+		this.pendingLayoutFlush = afterLayoutFlush( triggeredByScroll =>
```
`pendingLayoutFlush` can be an instance property of the component, like `checkScrollPositionHandler`. No need to recreate it on every `checkScrollAfterLayout` call.

Calling the same function returned by `afterLayoutFlush` has another advantage: `afterLayoutFlush` will not schedule another call if no layout was performed since the previous call:
```js
pendingLayoutFlush(); // schedules callback
pendingLayoutFlush(); // noop, already scheduled
pendingLayoutFlush(); // same
// browser processes animation frame and does layout
checkScrollPosition(); // called only once even thought requested 3 times
```

sgomes replied:
There's a complication in having `pendingLayoutFlush` as an instance property: unlike `throttle`, `afterLayoutFlush` generates a function that disregards any parameters, instead of passing them down to the wrapped function.

sgomes replied:
```javascript
const throttled = throttle( 
    param => console.log( 'throttle: ', param ), 100 );

throttled( 'hi there' );

const afterLayout = afterLayoutFlush( 
    param => console.log( 'afterLayoutFlush: ', param ) );

afterLayout( 'hi there' );
```

generates:

```
afterLayoutFlush:  undefined
throttle:  hi there
```

sgomes replied:
I added parameter passthrough to `afterLayoutFlush` and refactored the code based on your suggestions here. Thanks!

jsnajdr commented on client/components/infinite-scroll/index.js:
```diff
@@ -3,36 +3,118 @@
+
+	clearPending() {
+		if ( this.checkScrollPositionHandler ) {
+			this.checkScrollPositionHandler.cancel();
```
This is an instance property that always exists, no need to check.

sgomes replied:
Yep, that's right. Same for the new one too.

jsnajdr commented on client/components/infinite-scroll/index.js:
```diff
@@ -3,36 +3,118 @@
+			this.checkScrollPosition( triggeredByScroll )
+		);
+
+		this.pendingLayoutFlush();
```
This call should pass the `triggeredByScroll` value.

sgomes replied:
This code has been entirely reworked, with parameters now passing through.

jsnajdr commented on client/components/infinite-scroll/index.js:
```diff
@@ -3,36 +3,118 @@
+	static propTypes = propTypeDefinition;
 
 	componentDidMount() {
 		window.addEventListener( 'scroll', this.checkScrollPositionHandler );
```
The `checkScrollPositionHandler` expect a `triggeredByScroll` param that is `true`. In this case, it will receive the scroll even as parameter, which is also a truthy value, but still...

sgomes replied:
You're totally right, it's ugly. I created a new instance method for event handling, that always passes `true`.

sgomes commented:
@jsnajdr Hmm... I'm clearly not grokking the throttle and afterLayoutFlush methods. Sorry for all the silly mistakes. I'll be studying them a bit more in-depth today.

jsnajdr commented:
@sgomes To illustrate why cancelling a throttled function inside the callback is evil, try to run this Node.js script:
```js
const _ = require( 'lodash' );
const throttledFun = _.throttle( () => {
  throttledFun.cancel();
  console.log( 'called' );
}, 10000 );
setInterval( throttledFun, 1000 );
```
Although the function is throttled to run only once in 10 seconds, it will print "called" every second, as if no throttling was in place at all!

The sequence of events:
1. The first call to throttled function calls the inner function immediately (on "leading edge"), starts the 10s timer.
2. The inner function cancels the timer.
3. The second call to throttled function calls the inner function immediately again -- the previous internal state of `throttle` was thrown away, so we are again at step 1.

sgomes committed:
Add displayName to both versions of the component

sgomes committed:
Add support for wrapping parameters to afterLayoutFlush.

This allows us to pass parameters to the wrapped function.

sgomes committed:
Rework deferring logic in scroll-based InfiniteScroll after review.

sgomes commented:
Thank you so much for the clear and thorough explanation, @jsnajdr ! With that, I think I finally wrapped my head around all the indirectness, and realised I was really over-complicating things. The reworked code is much simpler, and hopefully correct now.

sgomes committed:
Remove deferred update on unmount in StickyPanel

jsnajdr reviewed (approved):
Thanks for all the hard work @sgomes! The patch works great and is ready to :ship: I left a few comments with opportunities for followup PRs. There are so many little improvements that can be done in this area.

sgomes commented on client/lib/after-layout-flush/index.js:
```diff
@@ -18,7 +18,7 @@ export default function afterLayoutFlush( func ) {
 	let timeoutHandle = undefined;
 
-	const scheduleRAF = rafFunc => () => {
+	const scheduleRAF = rafFunc => ( ...args ) => {
```
Good idea! Here it is: https://github.com/Automattic/wp-calypso/pull/26640
